### PR TITLE
chore(blockifier): impl traits Executable & TxInfoCreator for api_executabe_txs

### PR DESCRIPTION
**Pull request by avivg-starkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)

### CODE REVIEW

**Review by meship-starkware — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: 1 of 2 files reviewed, 2 unresolved discussions (waiting on @avivg-starkware and @noaov1)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 668 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O8pBZvf5qf7lXIaHWZv:-O8pBZvf5qf7lXIaHWZw:b-mio5os) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L668)):*
> ```XML
>                     // In the past, we allowed redeclaration of Cairo 0 contracts since there was
>                     // no class commitment (so no need to check if the class is already declared).
>                     state.set_contract_class(class_hash, contract_class(self)?)?;
> ```

Is this semi-column intentional? Also, it seems like all the functions are returning None upon success, so in which case are we returning CallInfo? Is it for future use? 
___
*[`crates/blockifier/src/transaction/transactions.rs` line 709 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O8pDftMAGxe9t8zYd3B:-O8pDftMAGxe9t8zYd3C:b-ultpwu) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L709)):*
> ```XML
>                     max_fee: tx.max_fee,
>                 })
>             }
> ```

Do they need to be in different cases because of the compiled class hash? It seems to me like it should work. 



_Suggestion:_
```XML
            starknet_api::transaction::DeclareTransaction::V0(tx)
            | starknet_api::transaction::DeclareTransaction::V1(tx) | 
            starknet_api::transaction::DeclareTransaction::V2(tx)  => {
                TransactionInfo::Deprecated(DeprecatedTransactionInfo {
                    common_fields,
                    max_fee: tx.max_fee,
                })
            }
```

**Review by meship-starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: 1 of 2 files reviewed, 2 unresolved discussions (waiting on @avivg-starkware and @noaov1)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 668 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O8pBZvf5qf7lXIaHWZv:-O8pwZ-B2CakdBWm8HeU:b-tzxxj9) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L668)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

Is this semi-column intentional? Also, it seems like all the functions are returning None upon success, so in which case are we returning CallInfo? Is it for future use?
</blockquote></details>

Never mind, I saw that it is part of this function API. Only the semicolon part is relevant

**Review by meship-starkware — COMMENTED** [review]
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: all files reviewed, 2 unresolved discussions (waiting on @avivg-starkware and @noaov1)

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: all files reviewed, 2 unresolved discussions (waiting on @avivg-starkware and @noaov1)

**Review by noaov1 — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: all files reviewed, 4 unresolved discussions (waiting on @avivg-starkware)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 10 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O94VzCFA2m5wV0lZdfg:-O94VzCFA2m5wV0lZdfh:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L10)):*
> ```XML
>     DeployAccountTransaction as StarknetApiDeployAccountTx,
>     InvokeTransaction as StarknetApiInvokeTx,
> };
> ```





_Suggestion:_
```XML
use starknet_api::executable_transaction::{
    DeclareTransaction as ExecutableDeclareTransaction,
    DeployAccountTransaction as ExecutableDeployAccountTx,
    InvokeTransaction as ExecutableInvokeTx,
};
```
___
*[`crates/blockifier/src/transaction/transactions.rs` line 693 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O94UELDCAMX_uHv_PRp:-O94UELE2tFUscBS3ART:bdjppj) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L693)):*
> ```XML
>             nonce: self.nonce(),
>             sender_address: self.sender_address(),
>             only_query: false, //QUESTION(AvivG): should get only_query from some source?
> ```

That is an interesting question. We can set it here to false and than override it by the account transaction that called this method.



_Code quote:_
```XML
            only_query: false, //QUESTION(AvivG): should get only_query from some source?
```

**Review by avivg-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @meship-starkware and @noaov1)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 10 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O94VzCFA2m5wV0lZdfg:-O9dOh9V4WSrpk0gvo0f:b-7cylbc) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L10)):*
> ```XML
>     DeployAccountTransaction as StarknetApiDeployAccountTx,
>     InvokeTransaction as StarknetApiInvokeTx,
> };
> ```

for consistency I've changed all to end with 'Tx' instead of  'Transaction'. Is there a reason you think 'ExecutableDeclareTransaction' should be different? or to have them all end with 'Transaction'?
___
*[`crates/blockifier/src/transaction/transactions.rs` line 693 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O94UELDCAMX_uHv_PRp:-O9dQgri026r3mmg6wUs:b-16zlpi) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L693)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

That is an interesting question. We can set it here to false and than override it by the account transaction that called this method.
</blockquote></details>

OK thank you. I left a temporary comment for myself to be removed before merging. 
___
*[`crates/blockifier/src/transaction/transactions.rs` line 709 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O8pDftMAGxe9t8zYd3B:-O9dZp8DC3oWKQ9MEkp9:bf9zsrm) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L709)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

Do they need to be in different cases because of the compiled class hash? It seems to me like it should work.
</blockquote></details>

'tx' in the first two is of type 'DeclareTransactionV0V1' and in the latter it is 'DeclareTransactionV2', because it is used to extract max_fee, it must be separated

**Review by noaov1 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-:-OAMv8ZW2uN2u0R4Z6SW:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @avivg-starkware and @meship-starkware)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 10 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-O94VzCFA2m5wV0lZdfg:-OAMrT-26qk5EJn6SKVW:bh7uleu) ([raw file](https://github.com/starkware-libs/sequencer/blob/f21ad3e9f274b562827c2d7e8f971807304f879d/crates/blockifier/src/transaction/transactions.rs#L10)):*
<details><summary><i>Previously, avivg-starkware wrote…</i></summary><blockquote>

for consistency I've changed all to end with 'Tx' instead of  'Transaction'. Is there a reason you think 'ExecutableDeclareTransaction' should be different? or to have them all end with 'Transaction'?
</blockquote></details>

No, I accidentally wrote it differently.

**Review by meship-starkware — APPROVED** [review]
> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @avivg-starkware)

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @avivg-starkware, @meship-starkware, and @noaov1)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 715 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwSYvO1vokA55ed8Pd:-OAwSYvO1vokA55ed8Pe:b-xs7o0p) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/blockifier/src/transaction/transactions.rs#L715)):*
> ```XML
>     ) -> TransactionExecutionResult<Option<CallInfo>> {
>         let class_hash = self.class_hash();
>         let ctor_context = ConstructorContext {
> ```

Avoid abbreviations. 



_Suggestion:_
```XML
let constructor_context = ConstructorContext {
```
___
*[`crates/blockifier/src/transaction/transactions.rs` line 852 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwTOcWBBPr6f5sAHVf:-OAwTOcWBBPr6f5sAHVg:by8lm26) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/blockifier/src/transaction/transactions.rs#L852)):*
> ```XML
>     class_hash: ClassHash,
>     compiled_class_hash: Option<CompiledClassHash>,
> ) -> TransactionExecutionResult<()> {
> ```

Document.



_Suggestion:_
```XML
/// Tries declaring a contract class i.e. setting the contract class at the state at the given class hash.
fn try_declare<S: State>(
    tx: &ExecutableDeclareTx,
    state: &mut S,
    class_hash: ClassHash,
    compiled_class_hash: Option<CompiledClassHash>,
) -> TransactionExecutionResult<()> {
```
___
*[`crates/blockifier/src/transaction/transactions.rs` line 852 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwUD8ZCr7e_Uf6Amu9:-OAwUD8ZCr7e_Uf6AmuA:bn1con5) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/blockifier/src/transaction/transactions.rs#L852)):*
> ```XML
>     class_hash: ClassHash,
>     compiled_class_hash: Option<CompiledClassHash>,
> ) -> TransactionExecutionResult<()> {
> ```

see that for the `Declare` struct in this file - `try_declare` is a method. There is code duplication here. Add a todo to delete it there (Or, can you please link the PR where the code duplication is resolved?)

Also, please do the same for other code duplications around this PR. 



_Code quote:_
```XML
fn try_declare<S: State>(
    tx: &ExecutableDeclareTx,
    state: &mut S,
    class_hash: ClassHash,
    compiled_class_hash: Option<CompiledClassHash>,
) -> TransactionExecutionResult<()> {
```
___
*[`crates/blockifier/src/transaction/transactions.rs` line 872 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwTvZU784CECv1z-Xj:-OAwTvZU784CECv1z-Xk:b-vxvwah) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/blockifier/src/transaction/transactions.rs#L872)):*
> ```XML
> fn contract_class(tx: &ExecutableDeclareTx) -> ContractClass {
>     tx.class_info.contract_class.clone()
> }
> ```

Can't this be a getter method of `ExecutableDeclareTx`? (i.e. update the code in Starknet API).



_Code quote:_
```XML
fn contract_class(tx: &ExecutableDeclareTx) -> ContractClass {
    tx.class_info.contract_class.clone()
}
```
___
*[`crates/starknet_api/src/executable_transaction.rs` line 232 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwRP4rCBHflbmTkXW1:-OAwRP4rCBHflbmTkXW2:b-gp164z) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/starknet_api/src/executable_transaction.rs#L232)):*
> ```XML
>         (signature, TransactionSignature),
>         (version, TransactionVersion)
>     );
> ```

TL;DR" move to top of `impl` block.


Align the order of calls with other `impl` blocks.

For `DeployAccountTransaction` and `InvokeTransaction` it is:
1. ` implement_inner_tx_getter_calls`,
2. `pub fn tx(&self)` (no need to add this function if it is never used).
3. `create`
4. `from_rpc_tx` (Irrelevant for declare - long story).



_Code quote:_
```XML
    implement_inner_tx_getter_calls!(
        (class_hash, ClassHash),
        (nonce, Nonce),
        (sender_address, ContractAddress),
        (signature, TransactionSignature),
        (version, TransactionVersion)
    );
```

**Review by avivg-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @ArniStarkware, @meship-starkware, and @noaov1)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 715 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwSYvO1vokA55ed8Pd:-OAwiAGl4BmY-asY-IxG:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/blockifier/src/transaction/transactions.rs#L715)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Avoid abbreviations.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/transactions.rs` line 852 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwTOcWBBPr6f5sAHVf:-OAwiLy--GcPNQrZdY2X:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/blockifier/src/transaction/transactions.rs#L852)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Document.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/transactions.rs` line 852 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwUD8ZCr7e_Uf6Amu9:-OAwicU32b9jlixm52E2:bpeuv3i) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/blockifier/src/transaction/transactions.rs#L852)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

see that for the `Declare` struct in this file - `try_declare` is a method. There is code duplication here. Add a todo to delete it there (Or, can you please link the PR where the code duplication is resolved?)

Also, please do the same for other code duplications around this PR.
</blockquote></details>

Yes, it is removed together with "DeclareTransaction" entire implementation in the following PR:
https://github.com/starkware-libs/sequencer/pull/1831
___
*[`crates/blockifier/src/transaction/transactions.rs` line 872 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwTvZU784CECv1z-Xj:-OAwk6If5y2tjg9rWsUR:b-tnn1sh) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/blockifier/src/transaction/transactions.rs#L872)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Can't this be a getter method of `ExecutableDeclareTx`? (i.e. update the code in Starknet API).
</blockquote></details>

It can. See change
___
*[`crates/starknet_api/src/executable_transaction.rs` line 232 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwRP4rCBHflbmTkXW1:-OAwkxsK3pURFtti7xTJ:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/f3a7c109c6a1ccd79e3e4dce60c308a2158097e9/crates/starknet_api/src/executable_transaction.rs#L232)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

TL;DR" move to top of `impl` block.


Align the order of calls with other `impl` blocks.

For `DeployAccountTransaction` and `InvokeTransaction` it is:
1. ` implement_inner_tx_getter_calls`,
2. `pub fn tx(&self)` (no need to add this function if it is never used).
3. `create`
4. `from_rpc_tx` (Irrelevant for declare - long story).
</blockquote></details>

Done.

**Comment by ArniStarkware** [comment]
*[`crates/blockifier/src/transaction/transactions.rs` line 858 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwofkq-Y2ebgk0Zu2Q:-OAwofkq-Y2ebgk0Zu2R:byijkqo) ([raw file](https://github.com/starkware-libs/sequencer/blob/dacda4c33e384edfeeba9e6c9af464ed0f91b67d/crates/blockifier/src/transaction/transactions.rs#L858)):*
> ```XML
>         Err(StateError::UndeclaredClassHash(_)) => {
>             // Class is undeclared; declare it.
>             state.set_contract_class(class_hash, tx.contract_class().try_into()?)?;
> ```

Note to self. This is where the magic happens!
Nice.



_Code quote:_
```XML
contract_class().try_into()
```

**Review by ArniStarkware — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-:-OAwon94DxruWDCd8V-l:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @meship-starkware and @noaov1)

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: all files reviewed, 1 unresolved discussion (waiting on @avivg-starkware, @meship-starkware, and @noaov1)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 848 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwpLJZ2c99Rdv7iE3K:-OAwpLJZ2c99Rdv7iE3L:b-bum23r) ([raw file](https://github.com/starkware-libs/sequencer/blob/dacda4c33e384edfeeba9e6c9af464ed0f91b67d/crates/blockifier/src/transaction/transactions.rs#L848)):*
> ```XML
> 
> /// Tries declaring a contract class i.e. setting the contract class at the state at the given class
> /// hash.
> ```


ChatGPT suggests:



_Suggestion:_
```XML
/// Attempts to declare a contract class by setting the contract class in the state with the specified class hash.
```

**Review by avivg-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: all files reviewed, 1 unresolved discussion (waiting on @ArniStarkware, @meship-starkware, and @noaov1)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 848 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/1279#-OAwpLJZ2c99Rdv7iE3K:-OAx8zx3DQB0sfq7uEL4:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/dacda4c33e384edfeeba9e6c9af464ed0f91b67d/crates/blockifier/src/transaction/transactions.rs#L848)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

ChatGPT suggests:
</blockquote></details>

Done.

**Review by ArniStarkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: 2 of 3 files reviewed, all discussions resolved (waiting on @meship-starkware and @noaov1)

**Review by ArniStarkware — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @meship-starkware and @noaov1)

**Review by meship-starkware — APPROVED** [review]
> Reviewed 1 of 2 files at r5, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @noaov1)

**Review by noaov1 — APPROVED** [review]
> Reviewed 1 of 2 files at r5, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @avivg-starkware)

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @avivg-starkware)

**Review by ArniStarkware — APPROVED** [review]
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @avivg-starkware)

**Review by meship-starkware — APPROVED** [review]
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1279)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @avivg-starkware)
